src/devices/Arduino/ArduinoBoard.cs

```diff
@@ -0,0 +1,421 @@
+        public ArduinoBoard(Stream serialPortStream)
+        {
+            _dataStream = serialPortStream;
+            _spiEnabled = 0;
+            _supportedPinConfigurations = new List<SupportedPinConfiguration>();
+            _firmwareName = string.Empty;
+        }
+
+        /// <summary>
+        /// Creates an instance of the Arduino board connection connected to a serial port
+        /// </summary>
+        /// <param name="portName">Port name. On Windows, this is usually "COM3" or "COM4" for an Arduino attached via USB.
+        /// On Linux, possible values include "/dev/ttyAMA0", "/dev/serial0", "/dev/ttyUSB1", etc.</param>
+        /// <param name="baudRate">Baudrate to use. It is recommended to use at least 115200 Baud.</param>
+        public ArduinoBoard(string portName, int baudRate)
+        {
+            _serialPort = new SerialPort(portName, baudRate);
+            _serialPort.Open();
+            _dataStream = _serialPort.BaseStream;
+            _supportedPinConfigurations = new List<SupportedPinConfiguration>();
+            _firmwareName = string.Empty;
+        }
+
+        /// <summary>
+        /// Default pin numbering scheme of this board. Always <see cref="PinNumberingScheme.Logical"/>, because the Arduino has only one pin numbering scheme.
+        /// </summary>
+        public PinNumberingScheme DefaultPinNumberingScheme
+        {
+            get
+            {
+                // We have only one scheme
+                return PinNumberingScheme.Logical;
+            }
+        }
+
+        /// <summary>
+        /// Attach to this event to retrieve log messages
+        /// </summary>
+        public event Action<string, Exception?>? LogMessages;
+
+        /// <summary>
+        /// Searches the given list of com ports for a firmata device.
+        /// </summary>
+        /// <param name="comPorts">List of com ports. See <see cref="GetSerialPortNames"/>.</param>
+        /// <param name="baudRates">List of baud rates to test. <see cref="CommonBaudRates"/>.</param>
+        /// <returns>A board, already open and initialized. Null if none was found.</returns>
+        public static ArduinoBoard? FindBoard(List<string> comPorts, List<int> baudRates)
+        {
+            foreach (var port in comPorts)
+            {
+                foreach (var baud in baudRates)
+                {
+                    ArduinoBoard? b = null;
+                    try
+                    {
+                        b = new ArduinoBoard(port, baud);
+                        b.Initialize();
+                        return b;
+                    }
+                    catch (Exception x) when (x is NotSupportedException || x is TimeoutException || x is IOException || x is UnauthorizedAccessException)
+                    {
+                        b?.Dispose();
+                    }
+                }
+            }
+
+            return null;
+        }
```

Comment:
you should not open all ports and check if they can initialize, you might accidentally cause some action in the unrelated device.

Reply:
True, there is a certain risk that this could happen, but how would you otherwise detect a board?

Reply:
Honestly, I'd not try to detect it over serial port (it kinda is the same scenario as if you tried to check all I2C addresses and checked if your device binding was able to connect to one of them)... The only way you can detect it reliably is if your device is USB showing as serial port and then getting some info from USB itself but not sure if that is worth the complexity...

Reply:
(with I2C you're likely to cause less likely to cause any harm though, with SerialPort there is almost nothing defining any universal format of data, most of the I2C at least uses registers and reads it more or less the same)

Reply:
Well, this is an optional feature (you do not have to use this auto-detect feature) and it works quite nicely for testing and you can specify the list of ports to scan (or ommit one that causes trouble). It is also very helpful since when connecting and disconnecting Arduinos (or generally USB-to-Serial devices) they often tend to change the port numbers. I don't want to change the code each time I reconnect a board.

Reply:
Added corresponding warning in documentation.